src/harness/sas_test_harness.py

```diff
@@ -0,0 +1,360 @@
+from urllib import quote
+from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
+
+CERT_FILE = 'certs/server.cert'
+KEY_FILE = 'certs/server.key'
+CA_CERT = 'certs/ca.cert'
+CIPHERS = [
+    'AES128-GCM-SHA256',              # TLS_RSA_WITH_AES_128_GCM_SHA256
+    'AES256-GCM-SHA384',              # TLS_RSA_WITH_AES_256_GCM_SHA384
+    'ECDHE-ECDSA-AES128-GCM-SHA256',  # TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+    'ECDHE-ECDSA-AES256-GCM-SHA384',  # TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+    'ECDHE-RSA-AES128-GCM-SHA256',    # TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+]
+
+class SasTestHarnessServer(threading.Thread):
+  """
+  SAS Test Harness class to create FAD records and to create a HTTP server
+  to serve the FAD records.This class inherits from the Python Thread class.
+  Every instance created for this class can be run under a separate thread. The
+  code to be executed by the thread is placed under the overridden method run ()
+  where a HTTP server is created to serves the FAD files.
+  """
+  def __init__(self, name, host_name, port, sas_version, dump_path, cert_file=None, key_file=None):
+    """ This constructor initializes the SasTestHarnessServer class with required information
+    passed as args.
+
+    Args:
+        name: A name for SAS Test Harness.If multiple SAS Test Hareness are created
+              then this value should be unique.
+        host_name: host name of SAS Test Harness that the SAS UUT should connect to for FAD records.
+        port:   port of SAS Test Harness host where http server is configured.
+        sas_version: sas version of the SAS Test.
+        dump_path: absolute file_path where the FAD records served by
+              the SAS Test Harness will be placed.
+        cert_file: The absolute path to the certificate used by SAS Test Harness
+              to authorize with SAS UUT.
+        key_file:  The absolute path to the private key file for the cert_file.
+    """
+    super(SasTestHarnessServer, self).__init__()
+    self.name = name
+    self.host_name = host_name
+    self.port = port
+    self.sas_version = sas_version
+
+    #BaseURL format should be https: // < hostname >: < port > / versionX.Y
+    self.http_server_url = "https://" + host_name + ':' + str(port) + '/' + sas_version
+    self.dump_path = dump_path
+    self.cert_file = cert_file if cert_file is not None else CERT_FILE
+    self.key_file = key_file if key_file is not None else KEY_FILE
+    self.server = None
+    self.setDaemon(True)
+
+  def __del__(self):
+     SasTestHarnessServer.cleanDumpFiles(self.dump_path)
+
+  def getBaseUrl(self):
+    return self.http_server_url
+
+  def getUrlHostNameAndPort(self):
+    """ This method will return the http_server_url used in startServer method by truncating other
+    querystring in url and returns tuple(<hostname>, <port>).
+    """
+    return (self.host_name, self.port)
+
+  def run(self):
+    """ Starts the HTTPServer as background thread.
+    """
```

Comment:
> The run() method is an overridden method from thread class and it gets invoked when the thread is started using thread.start().
>
> The start server code has been moved into this method to avoid confusion.


Please add a note like this to the run() and shutdown() methods so that it's clear to future readers of the code.

Reply:
updated accordingly